data/src/main/scala/ch.epfl.scala.index.data/bintray/DownloadPoms.scala

```diff
@@ -1,90 +1,130 @@
-    groupBy(_.sha1).                        // remove duplicates with sha1
-    map{case (_, vs) => vs.head}
-
-  private val progress = new ProgressBar("Downloading POMs", searchesBySha1.size)
-  
-  private def pomPath(search: BintraySearch) = bintrayPomBase.resolve(s"${search.sha1}.pom")
-  
-  private val downloadPoms = Source(searchesBySha1)
-    .map(s => (download(s), s))
-    .via(splitConnectionPool)
-    .mapAsync(1){
-      case (Success(response), search) => {
-        Unmarshal(response).to[String].map(body =>
-          if(response.status == StatusCodes.OK) Right((body, search))
-          else Left((body, search))
-        )
+
+  /**
+   * verify the sha1 file hash
+   * @param path the path to the file to verify
+   * @param sha1 the sha1 hash
+   * @return
+   */
+  private def verifySHA1FileHash(path: Path, sha1: String): Boolean = {
+
+    val source = scala.io.Source.fromFile(path.toFile)
+    val content = source.mkString.split(nl).mkString("")
+    source.close()
+
+    verifyChecksum(content, sha1)
+  }
+  /**
+   * get a list of bintraySearch object where no sha1 file exists
+   */
+  private val searchesBySha1: Set[BintraySearch] = {
+
+    BintrayMeta.readQueriedPoms(bintrayCheckpoint)
+      .filter(s => !Files.exists(pomPath(s)) || !verifySHA1FileHash(pomPath(s), s.sha1))
+      .groupBy(_.sha1)                         // remove duplicates with sha1
+      .map { case (_, vs) => vs.head }
+      .toSet
+  }
+
+  /**
+   * partly url encode - replaces only spaces
+   *
+   * @param url the url to "encode"
+   * @return
+   */
+  private def escape(url: String) = url.replace(" ", "%20")
+
+  /**
+   * get the download request object
+   * downloads pom from
+   * - jcenter.bintray.com or
+   * - dl.bintray.com
+   *
+   * @param search the bintray Search
+   * @return
+   */
+  private def downloadRequest(search: BintraySearch): WSRequest = {
+
+    if (search.repo == "jcenter" && search.owner == "bintray") {
+
+      wsClient.url(escape(s"https://jcenter.bintray.com/${search.path}"))
+    } else {
+
+      wsClient.url(escape(s"https://dl.bintray.com/${search.owner}/${search.repo}/${search.path}"))
+    }
+  }
+
+  /**
+   * handle the downloaded pom and write it to file
+   * @param search the bintray search
+   * @param response the download response
+   */
+  private def processPomDownload(search: BintraySearch, response: WSResponse): Unit = {
+
+    if (200 == response.status) {
+
+      val path = pomPath(search)
+
+      if (Files.exists(path)) {
+
+        Files.delete(path)
       }
-      case (Failure(e), _) => Future.failed(e)
-    }.map{
-      case Right((pom, search)) => {
-        progress.step()
-
-        val printer = new scala.xml.PrettyPrinter(80, 2)
-        val pw = new java.io.PrintWriter(pomPath(search).toFile)
-        pw.println(printer.format(scala.xml.XML.loadString(pom)))
-        pw.close
+
+      if (verifyChecksum(response.body, search.sha1)) {
```

Comment:
@MasseGuillaume: Here i check the Checksum - to make sure we have the right content. As mentioned earlier the writing is now native to make sure SHA1 hash will be the same. Do you see problems?